theseus/utils/utils.py

```diff
@@ -200,27 +201,67 @@ def autograd_fn(*optim_var_tensors):
 # with Timer("cuda:0") as timer:
 #    do_some_stuff()
 # print(timer.elapsed_time)
+##
+# timer = Timer("cuda:0")
+# timer.start()
+# do_some_stuff()
+# timer.end()
+# print(timer.elapsed_time)
+#
+# The timer can also optionally collect history of times in a dictionary
+# by adding caller ids to each context or start() calls.
+# For example,
+#
+# timer.start("f1")
+# f1()
+# timer.end()
+# with timer("f2"):
+#   f2()
+# print(timer.stats())
 class Timer:
-    def __init__(self, device: th.DeviceType) -> None:
+    def __init__(self, device: th.DeviceType, active: bool = True) -> None:
+        self.active = active
         self.device = torch.device(device)
         self.elapsed_time = 0.0
+        self._stats: Dict[str, List[float]] = defaultdict(list)
+        self._caller: Optional[str] = None
 
-    def __enter__(self) -> "Timer":
+    def start(self, caller: Optional[str] = None) -> "Timer":
+        if not self.active:
+            return self
         if self.device.type == "cuda":
             self._start_event = torch.cuda.Event(enable_timing=True)
             self._end_event = torch.cuda.Event(enable_timing=True)
             self._start_event.record()
         else:
             self._start_time = time.perf_counter_ns()
+        self._caller = caller
         return self
 
-    def __exit__(self, exc_type: Any, exc_value: Any, traceback: Any) -> None:
+    def end(self) -> None:
+        if not self.active:
+            return
         if self.device.type == "cuda":
             self._end_event.record()
             torch.cuda.synchronize()
             self.elapsed_time = self._start_event.elapsed_time(self._end_event) / 1e3
         else:
             self.elapsed_time = (time.perf_counter_ns() - self._start_time) / 1e9
+        if self._caller is not None:
+            self._stats[self._caller].append(self.elapsed_time)
+
+    def __call__(self, caller: Optional[str] = None) -> "Timer":
+        self.start(caller=caller)
+        return self
+
+    def __enter__(self) -> "Timer":
+        return self
```

Comment:
Should this be `self.start()`? otherwise does it the break the older behavior?

Reply:
Yeah, I fixed it in [b80f791](https://github.com/facebookresearch/theseus/pull/628/commits/b80f7915584d333cfef9ba306ac14c2b80956809). Good eye!

Reply:
Thanks for adding the unit tests as well!